### DISCUSSION TITLE
Advice on protecting API with auth token

### DISCUSSION THREAD

**Discussion by toychicken** [beginner-corp/begin-community#110] [General] [answered]
Hi all, 

I'm using the begin/create-react-app example as a template - but I've reached the point where I'd like to add an auth check on my /api calls - I'm using auth0 - and their recommended way seems to be to put a bearer token in the headers, and have the API endpoints verify the token... but whilst there's plenty of reference material for doing this with an express app, I can't find anything that might help with a begin.com app. Any pointers / articles etc would be really helpful.

Cheers

**Comment by tbeseda** [comment]
If you're trying to authenticate a request from your front end, @macdonst actually has a good article around auth0 + Begin/Architect. It doesn't use a front end framework like React, so you might need to pull out and implement a few things in a "React way".
https://blog.begin.com/posts/2022-01-20-auth0-todo-app
I like the auth middleware mentioned near the bottom. It uses the built in sessions feature with JWTs so you don't need to inspect headers.
Let me know if that helps or not 🥂

**Reply by toychicken** [reply]
Thanks @tbeseda - that's a really helpful starting point. I've got another article that deals with the react-y side of things, so I'll let you know how I get on! :)

**Comment by macdonst** [comment]
Hey, @toychicken are you putting a base64 encoded JWT into a header and need to decode and verify the bearer token is correct?

If so it's not far off from the way express.js does it but I'd be happy to answer the question and turn it into another blog post :)

**Reply by toychicken** [reply]
Yeah, that's where I am at the moment. I've got the jwt being passed in the header from the client app, but I'm flailing a bit with what to do at the other end. Pointers (and indeed a blog post) would be really helpful :)

**Reply by toychicken** [reply]
I'm assuming I can use the `arc.functions.http.async(auth, handler)` kind of approach, but... yeah falling over on how to handle the 'auth' part :)

**Comment by toychicken** [comment]
OK - so here's where I am right after a bit of playing around.
I have an Auth0 SPA application set up, and a react app that logs in / out, and gets the accessToken nicely. I then put that access token in the headers of requests to my API.
On the API side, I'm using `arc.http.async(auth, handler);` - and the auth function looks something like:

```
// src/shared/auth.js
const arc = require('@architect/functions')
const jwt_decode = require('jwt-decode')

module.exports = { auth }

async function auth(req) {

	const token = req.headers.authorization;

	if (!token) {
		return {
			status: 401,
			json: { errors: [ 'authorization_token_missing' ] }
		}
	}

	try {
		const decoded = jwt_decode(token)
		console.log('decoded', decoded)
		if (decoded.azp !== process.env.AUTH0_CLIENT_ID) {
			return {
				status: 401,
				json: { errors: [ 'token_audience_invalid' ] }
			}
		}
	} catch (err) {
		return {
			status: 401,
			json: { errors: [ 'invalid_jwt' ] }
		}
	}
	return
}
```

And this "works" - but I'm not confident that I should be validating by testing `decoded.azp` against the client_id. Is this right?

**Comment by macdonst** [comment, marked as answer]
@toychicken is your JWT signed with a private key or a secret? I recommend that you get Auth0 to return you a signed JWT that you can verify with the public key. You'll want to use the [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken) package, not jwt-decode package both from Auth0. The `jsonwebtoken` package has a `verify` function that will make sure the token is validated against your public/private key. Then it returns the decoded JWT that you can check to make sure the client ID is also correct.
